File: frontend/src/types/candidate.ts
export interface WorkExperience {
  id: string;
  company_name: string;
  job_title: string;
  employment_type: "full_time" | "part_time" | "contract" | "internship" | "freelance";
  location: string | null;
  is_current: boolean;
  start_date: string | null;
  end_date: string | null;
  description: string | null;
  achievements: string[] | null;
}

export interface Education {
  id: string;
  institution: string;
  degree: string | null;
  field_of_study: string | null;
  grade: string | null;
  start_year: number | null;
  end_year: number | null;
  is_current: boolean;
  description: string | null;
}

export interface Certification {
  id: string;
  name: string;
  issuing_org: string | null;
  issue_date: string | null;
  expiry_date: string | null;
  credential_id: string | null;
  credential_url: string | null;
}

export interface Project {
  id: string;
  title: string;
  description: string | null;
  tech_stack: string[] | null;
  project_url: string | null;
  repo_url: string | null;
  start_date: string | null;
  end_date: string | null;
}

export interface CandidateSkill {
  id: string;
  skill_name: string;
  proficiency: number;
  years_exp: number | null;
}

export interface CandidateProfile {
  id: string;
  user_id: string;
  full_name: string | null;
  phone: string | null;
  location: string | null;
  headline: string | null;
  summary: string | null;
  avatar_url: string | null;
  resume_url: string | null;
  resume_filename: string | null;
  desired_role: string | null;
  desired_salary_min: number | null;
  desired_salary_max: number | null;
  desired_location: string | null;
  open_to_remote: boolean;
  notice_period_days: number | null;
  years_of_experience: number | null;
  profile_strength: number;
  created_at: string;
  updated_at: string;
  work_experiences: WorkExperience[];
  educations: Education[];
  certifications: Certification[];
  projects: Project[];
  skills: CandidateSkill[];
}

export interface ProfileStrength {
  score: number;
  breakdown: Record<string, number>;
  tips: string[];
}

export interface SalaryBenchmark {
  role: string;
  location: string;
  min_salary: number;
  median_salary: number;
  max_salary: number;
  currency: string;
  source: string;
}

File: backend/app/schemas/candidate.py
"""Pydantic schemas for candidate profiles."""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.candidate import EmploymentType


class WorkExperienceCreate(BaseModel):
    company_name: str
    job_title: str
    employment_type: EmploymentType = EmploymentType.FULL_TIME
    location: str | None = None
    is_current: bool = False
    start_date: datetime | None = None
    end_date: datetime | None = None
    description: str | None = None
    achievements: list[str] | None = None


class WorkExperienceResponse(WorkExperienceCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class EducationCreate(BaseModel):
    institution: str
    degree: str | None = None
    field_of_study: str | None = None
    grade: str | None = None
    start_year: int | None = None
    end_year: int | None = None
    is_current: bool = False
    description: str | None = None


class EducationResponse(EducationCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class CertificationCreate(BaseModel):
    name: str
    issuing_org: str | None = None
    issue_date: datetime | None = None
    expiry_date: datetime | None = None
    credential_id: str | None = None
    credential_url: str | None = None


class CertificationResponse(CertificationCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class ProjectCreate(BaseModel):
    title: str
    description: str | None = None
    tech_stack: list[str] | None = None
    project_url: str | None = None
    repo_url: str | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None


class ProjectResponse(ProjectCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class CandidateSkillCreate(BaseModel):
    skill_name: str
    proficiency: int = Field(default=3, ge=1, le=5)
    years_exp: float | None = None


class CandidateSkillResponse(CandidateSkillCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class CandidateProfileUpdate(BaseModel):
    full_name: str | None = None
    phone: str | None = None
    location: str | None = None
    headline: str | None = None
    summary: str | None = None
    desired_role: str | None = None
    desired_salary_min: int | None = None
    desired_salary_max: int | None = None
    desired_location: str | None = None
    open_to_remote: bool | None = None
    notice_period_days: int | None = None
    years_of_experience: float | None = None


class CandidateProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    full_name: str | None
    phone: str | None
    location: str | None
    headline: str | None
    summary: str | None
    avatar_url: str | None
    resume_url: str | None
    resume_filename: str | None
    desired_role: str | None
    desired_salary_min: int | None
    desired_salary_max: int | None
    desired_location: str | None
    open_to_remote: bool
    notice_period_days: int | None
    years_of_experience: float | None
    profile_strength: int
    created_at: datetime
    updated_at: datetime
    work_experiences: list[WorkExperienceResponse] = []
    educations: list[EducationResponse] = []
    certifications: list[CertificationResponse] = []
    projects: list[ProjectResponse] = []
    skills: list[CandidateSkillResponse] = []


class ProfileStrengthResponse(BaseModel):
    score: int
    breakdown: dict[str, int]
    tips: list[str]
